multi asset enum

gilescope opened the pull request:
Allow swapping with native. (Ideally rather than this enum we could use just multilocation and a couple of convert functions)

gilescope committed:
multiasset enum

gilescope commented on frame/dex/src/lib.rs:
```diff
@@ -413,9 +424,22 @@ pub mod pallet {
+					MultiAssetId::Asset(asset2) => {
+						T::Assets::transfer(asset2, &pallet_account, &withdraw_to, amount2, false)?;
+					},
+				}
```
Could pull these out into a transfer function that takes a MultiAsset.

jsidorenko replied:
yeah, I think it's a good idea

gilescope commented:
While this passes all the tests, this currently relies on AssetBalance and Currency Balance being the same type. (This is the case for statemine/t fyi)

gilescope committed:
only allow native currency pairs

gilescope committed:
added slippage tests

jsidorenko commented on frame/dex/src/lib.rs:
```diff
@@ -307,8 +311,22 @@ pub mod pallet {
 				let pallet_account = Self::account_id();
-				T::Assets::transfer(asset1, &sender, &pallet_account, amount1, false)?;
-				T::Assets::transfer(asset2, &sender, &pallet_account, amount2, false)?;
+				match asset1 {
+					MultiAssetId::Native => {
+						T::Currency::transfer(&sender, &pallet_account, amount1, false)?;
+					},
+					MultiAssetId::Asset(asset1) => {
+						T::Assets::transfer(asset1, &sender, &pallet_account, amount1, false)?;
+					},
+				}
```
I would create a helper transfer method for that

gilescope committed:
transfer into separate method

(Also fee not set in 2 places now.)

Added test where lp and user are different users.
